elastic/beats
# Implement k8s secrets provider for Agent

This PR aims to add support for consuming k8s secrets through Agent's variables. 

Note: This is still a draft but comments are needed so as to verify it's going to the right direction, so plz @blakerouse @exekias feel free to comment here.

How to test this:

0. Create the k8s secret:
```
cat << EOF | kubectl apply -f -apiVersion: v1
kind: Secret
metadata:
  name: somesecret
type: Opaque
data:
  value: $(echo -n "passpass" | base64)
EOF
```
1. Use the following config:
```
providers.kubernetes_secrets:
  kube_config: /Users/chrismark/.kube/config

inputs:
  - type: logfile
    streams:
      - paths: ${env.logpath}/${kubernetes_secrets.default.somesecret.value}/another.log
```
2. Run `inspect` command to check what is the compiled configuration (use proper path for the `elastic-agent.yml` config file :
`logpath=/var/log ./elastic-agent -c /Users/ubuntu/Desktop/elastic-agent.yml inspect output -o default `

Verify that the following is produced:
```
[default] filebeat:
filebeat:
  inputs:
  - index: logs-generic-default
    paths: /var/log/passpass/another.log
```

Unit testing: 
1. `go test -v ./pkg/agent/transpiler/... -run TestVars_ReplaceWithFetchContextProvider`
2. `go test -v ./pkg/composable/providers/kubernetessecrets...`

Closes https://github.com/elastic/beats/issues/24143
